Sources/Signal.swift

```diff
@@ -1492,6 +1492,83 @@ extension SignalProtocol {
 		}
 	}
 
+	/// Conditionally throttles values sent on the receiver whenever
+	/// `shouldThrottle` is true, forwarding values on the given scheduler.
+	///
+	/// - note: While `shouldThrottle` remains false, values are forwarded on the
+	///         given scheduler.
+	///
+	/// - note: If multiple values are received while `shouldThrottle` is true,
+	///         the latest value is the one that will be passed on.
+	///
+	/// - note: If the input signal terminates while a value is being throttled,
+	///         that value will be discarded and the returned signal will
+	///         terminate immediately.
+	///
+	/// - parameters:
+	///   - shouldThrottle: A boolean property that controls whether values
+	///                     should be throttled.
+	///   - scheduler: A scheduler to deliver events on.
+	///
+	/// - returns: A signal that sends values only while `shouldThrottle` is false.
+	func throttle<P: PropertyProtocol>(while shouldThrottle: P, on scheduler: SchedulerProtocol) -> Signal<Value, Error>
+		where P.Value == Bool
+	{
+		return Signal { observer in
+			let initial: ThrottleWhileState<Value> = shouldThrottle.value ? .throttled(nil) : .resumed
+			let state = Atomic(initial)
+			let schedulerDisposable = SerialDisposable()
+
+			let disposable = CompositeDisposable()
+			disposable += schedulerDisposable
+			disposable += { _ = shouldThrottle }
+
+			disposable += shouldThrottle.producer.uniqueValues { $0 }.startWithValues { shouldThrottle in
+				state.modify { state -> ThrottleWhileState<Value> in
+					guard !state.isTerminated else { return state }
+
+					if shouldThrottle {
+						return .throttled(nil)
+					} else {
+						if case .throttled(let value?) = state {
+							schedulerDisposable.innerDisposable = scheduler.schedule {
+								observer.send(value: value)
```

Comment:
I think this value should get sent outside of the `Atomic`'s lock to prevent deadlock.

Even though it's in a `schedule` block, that can execute synchronously if, e.g., you're on the main queue and you're using the `UIScheduler`.

Same goes for the send below.